### PR TITLE
feat: handling attribute with null value

### PR DESCRIPTION
**Pull request by gpoujol**

### CODE REVIEW

File: nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java [hunk 1]
```diff
@@ -107,11 +107,11 @@ public ArrayList<Entity> parseNgsiLdEntities(JSONArray content) {
                         for (int j = 0; j < values.length(); j++) {
                             JSONObject value = values.getJSONObject(j);
                             AttributesLD attributesLD = parseNgsiLdAttribute(key, value);
-                            attributes.add(attributesLD);
+                            if (attributesLD.getAttrValue() !=null && attributesLD.getAttrValue().toString() != "null") attributes.add(attributesLD);
```
Comment (bobeal) [inline nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java]: instead of copy / pasting this line in different place, it could be handled in the two `parseNgsiLdXXXAttribute` methods

File: nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java [hunk 2]
```diff
@@ -166,7 +166,6 @@ private AttributesLD parseNgsiLdAttribute(String key, JSONObject value) {
             if (("Property".equals(attrType) && "unitCode".equals(keyOne))) {
                 if (value.get(keyOne) instanceof String)
                     subAttributes.add(new AttributesLD(keyOne.toLowerCase(), "Property", "", "", "", "", value.getString(keyOne), false, null));
-                else subAttributes.add(new AttributesLD(keyOne.toLowerCase(), "Property", "", "", "", "", null, false, null));
```
Comment (bobeal) [inline nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java:169]: why did you remove this line?

File: nifi-ngsi-bundle/nifi-ngsi-processors/src/test/resources/temporalEntities.json [hunk 3]
```diff
@@ -30,6 +30,11 @@
         }
       }
     },
+    "nullValue" : {
+      "type" : "Property",
+      "value" : null
+    },
+    "nullAttribute" : null,
```
Comment (bobeal) [inline nifi-ngsi-bundle/nifi-ngsi-processors/src/test/resources/temporalEntities.json:37]: maybe add a related check in the test?

Comment (gpoujol) [reply nifi-ngsi-bundle/nifi-ngsi-processors/src/test/resources/temporalEntities.json:37]: Yes I had this

File: nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java (hunk 2 above)
Comment (gpoujol) [reply nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java:169]: This line add a subAttribute with anull value but that is not what we want

File: nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java (hunk 1 above)
Comment (gpoujol) [reply nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java]: Ok I do this

File: nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java [hunk 4]
```diff
@@ -224,4 +223,11 @@ private AttributesLD parseNgsiLdSubAttribute(String key, JSONObject value) {
 
         return new AttributesLD(key.toLowerCase(), subAttrType, "", "", "", "", subAttrValue, false, null);
     }
+
+    //In NGSI-LD, we can't have a property with a null value or a null attribute
+    // So we do not add the attribute if it contains a null value or attribute is null
+    private void addAttributesIfCheked(ArrayList<AttributesLD> attributesLd, AttributesLD attributeLD) {
```
Comment (bobeal) [inline nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java]: It is `checked`, and not `cheked`. And `Valid` may be better.

File: nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java [hunk 5]
```diff
@@ -224,4 +223,11 @@ private AttributesLD parseNgsiLdSubAttribute(String key, JSONObject value) {
 
         return new AttributesLD(key.toLowerCase(), subAttrType, "", "", "", "", subAttrValue, false, null);
     }
+
+    //In NGSI-LD, we can't have a property with a null value or a null attribute
+    // So we do not add the attribute if it contains a null value or attribute is null
```
Comment (bobeal) [inline nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java]: Add a line (or more) to explain **why** we may have this situation. This is what is really important here. As a reader, I want to know why we have a null attribute (I can see by myself that you don't add it in this case)

File: nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java [hunk 6]
```diff
@@ -224,4 +223,14 @@ private AttributesLD parseNgsiLdSubAttribute(String key, JSONObject value) {
 
         return new AttributesLD(key.toLowerCase(), subAttrType, "", "", "", "", subAttrValue, false, null);
     }
+
+    //When I use this processor in a flow with 'Join Enrichment', this processor harmonize json. It can add attributes which are not present by default
+    //So this attributes is null. It can add also a null value.
+    //Moreover, when we do a temporal request, if attributes have no temporal values it is still added and it is null
+    //But in NGSI-LD, we can't have a property with a null value or a null attribute
+    //So we do not add the attribute if it contains a null value or attribute is null
+    private void addAttributesIfValid(ArrayList<AttributesLD> attributesLd, AttributesLD attributeLD) {
```
Comment (bobeal) [inline nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java]: you are adding just one attribute, so `addAttributeIfValid` is more accurate

File: nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java [hunk 7]
```diff
@@ -224,4 +223,14 @@ private AttributesLD parseNgsiLdSubAttribute(String key, JSONObject value) {
 
         return new AttributesLD(key.toLowerCase(), subAttrType, "", "", "", "", subAttrValue, false, null);
     }
+
+    //When I use this processor in a flow with 'Join Enrichment', this processor harmonize json. It can add attributes which are not present by default
+    //So this attributes is null. It can add also a null value.
+    //Moreover, when we do a temporal request, if attributes have no temporal values it is still added and it is null
+    //But in NGSI-LD, we can't have a property with a null value or a null attribute
+    //So we do not add the attribute if it contains a null value or attribute is null
```
Comment (bobeal) [inline nifi-ngsi-bundle/nifi-ngsi-processors/src/main/java/org/apache/nifi/processors/ngsi/ngsi/utils/NGSIUtils.java]:
```suggestion
    // When this processor is used in a flow with a `Join Enrichment` processor, it harmonizes JSON among all processed entities, for instance adding attributes which are not present by default in an entity.
    // In this case, these attributes are null or can have a null value.
    // Moreover, when doing a temporal request, if some attributes have no temporal values, they are still added and they are null
    // So we filter out attributes that contain a null value or whose whole value is null
```

**Review by bobeal — APPROVED** [review]
